runners/spark/src/main/java/org/apache/beam/runners/spark/translation/EvaluationContext.java

```diff
@@ -141,18 +140,15 @@ public void putDataset(PValue pvalue, Dataset dataset) {
     datasets.put(getOutput(transform), new UnboundedDataset<>(values, jssc, coder));
   }
 
-  public Dataset borrowDataset(PTransform<? extends PValue, ?> transform) {
-    return borrowDataset(getInput(transform));
+  public Dataset borrowDataset(PTransform<? extends PValue, ?> transform, boolean cacheHint) {
+    return borrowDataset(getInput(transform), cacheHint);
   }
 
-  public Dataset borrowDataset(PValue pvalue) {
+  public Dataset borrowDataset(PValue pvalue, boolean cacheHint) {
     Dataset dataset = datasets.get(pvalue);
     leaves.remove(dataset);
-    if (multiReads.contains(pvalue)) {
-      // Ensure the RDD is marked as cached
+    if (cacheHint) {
```

Comment:
Unless I'm getting really confused here (it's been a while since this was discussed) - we want to cache on `EvaluationContext#putDataset`, no ?
The new solution pre-visits to find which `RDD` to cache upon creation.
The old solution via `multiReads` was passive, and so if you already accessed this `RDD` you'll cache it.
So I thought we're going for: 
"I know this RDD is going to be consumed more than once, so I'll cache it now".

Makes sense ?